### PR TITLE
structuredClone/postMessage: add fast path for bare primitive values

### PR DESCRIPTION
**Pull request by robobun**
## What

`port.postMessage(1)`, `structuredClone(42)`, and the like fell off the structured-clone fast path into the full `CloneSerializer`. This adds `FastPath::Primitive` so a bare non-cell `JSValue` (int32, double, boolean, `null`, `undefined`, and `BigInt32` where enabled) is carried as-is.

## Why

The `canUseFastPath` block in `SerializedScriptValue::create` only looked at `value.isCell()` candidates (strings, `JSArray`, flat objects). A non-cell value fell past the block into the full serializer: transfer-list scan, `ObjectPool`/`HashSet` setup, `Vector<uint8_t>` heap alloc, version header + tag writes, plus a symmetric `CloneDeserializer` on receive.

A non-cell `JSValue` has no heap identity and is safe to carry across threads as-is. The existing `SimpleObject` fast path already stores primitive property values as bare cross-thread `JSValue`s, so this follows the same established pattern for the top-level value.

## How

- `SerializedScriptValue.h`: add `FastPath::Primitive`, `m_fastPathPrimitive`, `createPrimitiveFastPath`, and the matching private ctor.
- `SerializedScriptValue.cpp`: inside `if (canUseFastPath)` and before the `value.isCell()` branch, `if (!value.isCell()) return createPrimitiveFastPath(value);`. Ordering is load-bearing: on `JSVALUE64` the empty `JSValue` satisfies `isCell() == true` and keeps going down the existing branch; `SerializationForStorage` callers (`bun:jsc` `serialize()`) are already excluded by `canUseFastPath` and keep producing real bytes. Add the `Primitive` arm to both the `computeMemoryCost` and `deserialize` switches.
- `StructuredClone.cpp`: when the transfer list is empty and the value is non-cell, return it directly without constructing a `SerializedScriptValue`.

## Measurement

Release build, 1M iterations of `structuredClone(42)`:

| | before | after |
| --- | --- | --- |
| `structuredClone(42)` | ~532 ns/op | identity return |
| ratio vs `structuredClone(new Map())` (full serializer) | 0.52-0.78 | 0.005 |

## Testing

Added to `test/js/web/structured-clone-fastpath.test.ts`:

- Round-trip of every non-cell primitive (int32, doubles incl. `-0`/`NaN`/`Infinity`, booleans, `null`, `undefined`, small/large/negative BigInt) via `structuredClone`, `MessageChannel`, and a cross-thread `Worker` echo.
- `bun:jsc` `serialize()` still produces real bytes for every primitive (the `SerializationForStorage::Yes` path is unchanged).
- `structuredClone(42, { transfer: [buf] })` still detaches the buffer.
- `structuredClone(42)` runs at <0.25x the cost of `structuredClone(new Map())` (fails at ~0.5x on the previous build, passes at ~0.005x here).

Regression-checked against `test/js/web/workers/{structured-clone,structuredClone-classes,message-channel,message-port-pipe,worker-postmessage-transfer}.test.ts`, `test/js/web/broadcastchannel/broadcast-channel.test.ts`, and `test/js/node/worker_threads/worker_threads.test.ts` (553 tests, all pass).

---

**[review]** gate passed · iteration 0 · 4 files touched

<details><summary>fails on main (without fix)</summary>

```console
ASAN without fix: 1 FAILED
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/mechgate.xml" test/js/web/structured-clone-fastpath.test.ts
bun test v1.4.0 (cf9046958)

test/js/web/structured-clone-fastpath.test.ts:
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32) round-trips [5.71ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 negative) round-trips [0.81ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 zero) round-trips [0.46ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 max) round-trips [1.02ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 min) round-trips [0.44ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double) round-trips [0.46ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double -0) round-trips [0.54ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double NaN) round-trips [0.52ms]
(pass) Structured Clone Fast Path > bare primitive values > struc
... (truncated)

release without fix: 1 FAILED
bun test v1.4.0-canary.1 (1498d7b77)

test/js/web/structured-clone-fastpath.test.ts:
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32) round-trips [0.14ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 negative) round-trips [0.01ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 zero) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 max) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 min) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double -0) round-trips [0.01ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double NaN) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double Infinity) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double -Infinity) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structure
... (truncated)
```

</details>

<details><summary>passes on PR (with fix)</summary>

```console
ASAN with fix: all passed
$ BUN_DEBUG_QUIET_LOGS=1 bun scripts/build.ts --profile=debug --quiet test "--reporter=junit" "--reporter-outfile=/tmp/mechgate.xml" test/js/web/structured-clone-fastpath.test.ts
bun test v1.4.0 (cf9046958)

test/js/web/structured-clone-fastpath.test.ts:
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32) round-trips [5.25ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 negative) round-trips [0.58ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 zero) round-trips [0.34ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 max) round-trips [0.29ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 min) round-trips [0.32ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double) round-trips [0.30ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double -0) round-trips [0.37ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(double NaN) round-trips [0.35ms]
(pass) Structured Clone Fast Path > bare primitive values > struc
... (truncated)

release with fix: all passed
$ bun scripts/build.ts --profile=release
[configured] bun-profile → bun (stripped) in 969ms (unchanged)
ninja: Entering directory `/workspace/bun/build/release'
[1/27] gen cpp.rs (cppbind)
[1/27] cargo bun_bin → libbun_rust.a (--target x86_64-unknown-linux-gnu)

  nightly-2026-07-20-x86_64-unknown-linux-gnu unchanged - rustc 1.99.0-nightly (9f36de775 2026-07-19)

[1m[92m   Compiling[0m bun_bin v0.0.0 (/workspace/bun/src/bun_bin)
[1m[92m    Finished[0m `release` profile [optimized + debuginfo] target(s) in 4m 40s
[23/27] cxx obj/unified/UnifiedSource-src_jsc_bindings_webcore-0.cpp.o
[24/27] link bun-profile
[26/27] strip bun
[26/27] bun-profile --revision
1.4.0-canary.1+cf9046958
[build] done
bun test v1.4.0-canary.1 (cf9046958)

test/js/web/structured-clone-fastpath.test.ts:
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32) round-trips [0.10ms]
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 negative) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int32 zero) round-trips
(pass) Structured Clone Fast Path > bare primitive values > structuredClone(int3
... (truncated)
```

</details>

<details><summary>diff hotspot</summary>

```
src/jsc/bindings/webcore/SerializedScriptValue.cpp |  50 +++++---
 src/jsc/bindings/webcore/SerializedScriptValue.h   |   8 ++
 src/jsc/bindings/webcore/StructuredClone.cpp       |   4 +
 test/js/web/structured-clone-fastpath.test.ts      | 138 +++++++++++++++++++++
 4 files changed, 186 insertions(+), 14 deletions(-)
```

</details>

**gate history** · 1 passed · 0 rejected · iteration 0

<details><summary>evidence per changed file</summary>

```
file                                                reads  edits  tests
src/jsc/bindings/webcore/SerializedScriptValue.cpp      7      8      0
src/jsc/bindings/webcore/SerializedScriptValue.h        2      4      0
src/jsc/bindings/webcore/StructuredClone.cpp            4      2      0
test/js/web/structured-clone-fastpath.test.ts           2      9      0
```

</details>

### CODE REVIEW

File: src/jsc/bindings/webcore/SerializedScriptValue.cpp [hunk 1]
```diff
@@ -4446,6 +4456,15 @@ ExceptionOr<Ref<SerializedScriptValue>> SerializedScriptValue::create(JSGlobalOb
         && messagePorts.isEmpty();
 
     if (canUseFastPath) {
+        // A bare non-cell JSValue (int32/double/bool/null/undefined, plus BigInt32 when
+        // USE(BIGINT32)) has no heap identity and is safe to carry across threads as-is.
+        // Ordering is load-bearing: on JSVALUE64 the empty JSValue satisfies isCell()==true,
+        // so it continues to the cell branch below; this check must not be rewritten as
+        // isPrimitive() or hoisted above canUseFastPath (SerializationForStorage callers
+        // like bun:jsc serialize() depend on getting real wire bytes).
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/webcore/SerializedScriptValue.cpp]: Trimmed in 4b66cb6.

File: src/jsc/bindings/webcore/StructuredClone.cpp [hunk 2]
```diff
@@ -56,6 +56,11 @@ JSC_DEFINE_HOST_FUNCTION(jsFunctionStructuredClone, (JSC::JSGlobalObject * globa
     auto serializeOptions = convertDictionary<StructuredSerializeOptions>(*globalObject, options);
     RETURN_IF_EXCEPTION(throwScope, {});
 
+    // A non-cell JSValue (int32/double/bool/null/undefined) has no heap identity and structured
+    // clone is the identity function on it. With nothing to transfer, skip the round-trip.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/webcore/StructuredClone.cpp]: Trimmed in 4b66cb6.

File: test/js/web/structured-clone-fastpath.test.ts [hunk 3]
```diff
@@ -1,7 +1,136 @@
 import { describe, expect, test } from "bun:test";
 import { isASAN, rss } from "harness";
+import { serialize, deserialize } from "bun:jsc";
 
 describe("Structured Clone Fast Path", () => {
+  // === Primitive fast path tests ===
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/structured-clone-fastpath.test.ts:6]: The banner matches the existing section style in this file (`// === Array fast path tests ===` etc.). Perf-test comment trimmed to the durable invariant in 4b66cb6.

File: test/js/web/structured-clone-fastpath.test.ts [hunk 4]
```diff
@@ -1,7 +1,136 @@
...
+      ["double -Infinity", -Infinity],
+      ["double max safe int", Number.MAX_SAFE_INTEGER],
+      ["true", true],
+      ["false", false],
+      ["null", null],
+      ["undefined", undefined],
+      ["small BigInt", 1n],
+      ["large BigInt", 2n ** 80n],
+      ["negative BigInt", -(2n ** 80n)],
+    ];
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/structured-clone-fastpath.test.ts:32]: Added INT32_MIN/MAX, INT32_MAX+1, and the BigInt32 boundary values in 4b66cb6.

File: test/js/web/structured-clone-fastpath.test.ts [hunk 5]
```diff
@@ -1,7 +1,136 @@
...
+      if (typeof value === "number" && Number.isNaN(value)) {
+        expect(result).toBeNaN();
+      } else if (Object.is(value, -0)) {
+        expect(Object.is(result, -0)).toBe(true);
+      } else {
+        expect(result).toBe(value);
+      }
+      port1.close();
+      port2.close();
+    });
```
[the remark this answers is not in the document]
Comment (robobun) [reply test/js/web/structured-clone-fastpath.test.ts:64]: Done in 4b66cb6: `onmessageerror` rejects the promise, ports close in `finally`, and the worker loop wires `onmessageerror` to reject as well.

File: src/jsc/bindings/webcore/SerializedScriptValue.cpp [hunk 6]
```diff
@@ -4446,6 +4456,11 @@ ExceptionOr<Ref<SerializedScriptValue>> SerializedScriptValue::create(JSGlobalOb
         && messagePorts.isEmpty();
 
     if (canUseFastPath) {
+        // Not isPrimitive(): on JSVALUE64 the empty JSValue has isCell()==true and must keep
+        // falling through to the full serializer below.
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/webcore/SerializedScriptValue.cpp:4460]: Keeping this one. It is two short lines documenting a non-obvious JSC encoding detail (the empty `JSValue` satisfies `isCell()` on JSVALUE64) that would otherwise take a trip through `JSCJSValue.h` to rediscover, and it is what stops someone from rewriting the guard as `isPrimitive()`.

File: src/jsc/bindings/webcore/SerializedScriptValue.cpp [hunk 7]
```diff
@@ -4446,6 +4456,11 @@ ExceptionOr<Ref<SerializedScriptValue>> SerializedScriptValue::create(JSGlobalOb
...
+        if (!value.isCell())
+            return SerializedScriptValue::createPrimitiveFastPath(value);
```
[the remark this answers is not in the document]
Comment (robobun) [reply src/jsc/bindings/webcore/SerializedScriptValue.cpp:4462]: Done in cf90469: dropped the wrapper and outdented the body.
